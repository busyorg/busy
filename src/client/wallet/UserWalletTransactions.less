@import (reference) "../styles/custom.less";
@import (reference) "../styles/modules/variables.less";

.UserWalletTransactions {
  background-color: @white;
  border-radius: @border-radius-base;
  margin: 20px 0;

  &:empty {
    display: none;
  }

  &__transaction {
    color: @grey-nightRider;
    display: flex;
    padding: 10px 15px;
    border-right: 1px solid @white-gainsboro;
    border-left: 1px solid @white-gainsboro;
    border-bottom: 1px solid @white-gainsboro;

    &:nth-of-type(2) {
      border-top: 1px solid @white-gainsboro;
      border-top-left-radius: @border-radius-base;
      border-top-right-radius: @border-radius-base;
    }

    &:last-of-type {
      border-bottom-left-radius: @border-radius-base;
      border-bottom-right-radius: @border-radius-base;
    }
  }

  &__memo {
    word-wrap: break-word;
  }

  &__avatar {
    margin-top: 5px;
  }

  &__icon-container {
    align-items: center;
    color: @blue-heather;
    background-color: @blue-solitude;
    border-radius: 4px;
    display: flex;
    height: 40px;
    justify-content: center;
    margin-top: 4px;
    width: 44px;
  }

  &__icon {
    font-size: 22px !important;
  }

  &__content {
    display: flex;
    flex-direction: column;
    padding-left: 10px;
    width: 100%;
    overflow: hidden;
  }

  &__content-recipient {
    display: flex;
    flex-direction: row;
    font-weight: bold;
    white-space: nowrap;
  }

  &__timestamp {
    color: @blue-botticelli;
    font-size: 14px;
  }

  &__transfer {
    color: @red;
    margin-left: auto;
  }

  &__received {
    color: @green;
    margin-left: auto;
  }

  &__payout {
    margin-left: auto;
  }

  &__payout-rewards {
    &:after {
      content: ' - ';
      color: @blue-botticelli;
    }

    &:last-of-type {
      &:after {
        content: '';
      }
    }
  }

  @media @small-max-width {
    border-left: 0;
    border-right: 0;
    border-radius: 0;
  }

  @media @very-small-max-width {
    &__content-recipient {
      flex-wrap: wrap;
    }

    &__transfer {
      width: 100%;
      margin-left: 0;
    }

    &__received {
      width: 100%;
      margin-left: 0;
    }

    &__payout {
      width: 100%;
      margin-left: 0;
    }
  }
}
