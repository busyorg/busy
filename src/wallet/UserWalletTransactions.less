@import (reference) "../styles/custom.less";
@import (reference) "../styles/modules/variables.less";

.UserWalletTransactions {
  background-color: @white;
  border: 1px solid @white-gainsboro;
  border-radius: @border-radius-base;
  margin: 20px 0;

  &:empty {
    display: none;
  }

  &__transaction {
    color: @grey-nightRider;
    display: flex;
    padding: 10px 0;
    margin: 0 15px;
    border-bottom: 1px solid @white-gainsboro;

    &:last-of-type {
      border-bottom: 0;
    }
  }

  &__memo {
    word-wrap: break-word;
  }

  &__avatar {
    margin-top: 5px;
  }

  &__icon-container {
    align-items: center;
    color: @blue-heather;
    background-color: @blue-solitude;
    border-radius: 4px;
    display: flex;
    height: 40px;
    justify-content: center;
    margin-top: 4px;
    width: 44px;
  }

  &__icon {
    font-size: 22px !important;
  }

  &__content {
    display: flex;
    flex-direction: column;
    padding-left: 10px;
    width: 100%;
    overflow: hidden;
  }

  &__content-recipient {
    display: flex;
    flex-direction: row;
    font-weight: bold;
    white-space: nowrap;
  }

  &__timestamp {
    color: @blue-botticelli;
    font-size: 14px;
  }

  &__transfer {
    color: @red;
    margin-left: auto;
  }

  &__received {
    color: @green;
    margin-left: auto;
  }

  &__payout {
    margin-left: auto;
  }

  &__payout-rewards {
    &:after {
      content: ' - ';
      color: @blue-botticelli;
    }

    &:last-of-type {
      &:after {
        content: '';
      }
    }
  }

  @media @small-max-width {
    border-left: 0;
    border-right: 0;
    border-radius: 0;
  }

  @media @very-small-max-width {
    &__content-recipient {
      flex-wrap: wrap;
    }

    &__transfer {
      width: 100%;
      margin-left: 0;
    }

    &__received {
      width: 100%;
      margin-left: 0;
    }

    &__payout {
      width: 100%;
      margin-left: 0;
    }
  }
}
