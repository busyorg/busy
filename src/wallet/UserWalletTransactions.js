import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { FormattedNumber } from 'react-intl';
import ReduxInfiniteScroll from 'redux-infinite-scroll';
import { defaultAccountLimit } from '../helpers/apiHelpers';
import Loading from '../components/Icon/Loading';
import ReceiveTransaction from './ReceiveTransaction';
import TransferTransaction from './TransferTransaction';
import SavingsTransaction from './SavingsTransaction';
import PowerUpTransaction from './PowerUpTransaction';
import ClaimReward from './ClaimReward';
import './UserWalletTransactions.less';

class UserWalletTransactions extends React.Component {
  static propTypes = {
    transactions: PropTypes.arrayOf(PropTypes.shape()),
    getMoreUserAccountHistory: PropTypes.func.isRequired,
    currentUsername: PropTypes.string,
    totalVestingShares: PropTypes.string.isRequired,
    totalVestingFundSteem: PropTypes.string.isRequired,
  };

  static defaultProps = {
    transactions: [],
    currentUsername: '',
  };

  constructor(props) {
    super(props);
    const firstTransaction = _.head(props.transactions);
    this.state = {
      userHasMoreTransactions: firstTransaction.actionCount > defaultAccountLimit,
      lastActionCount: firstTransaction ? firstTransaction.actionCount : 0,
      loadingMoreTransactions: false,
    };
  }

  getFormattedTransactionAmount = (amount, currency) => {
    if (_.isEmpty(amount)) {
      return null;
    }

    const transaction = amount.split(' ');
    const transactionAmount = parseFloat(transaction[0]);
    const transactionCurrency = currency || transaction[1];
    return (
      <span>
        <FormattedNumber
          value={transactionAmount}
          minimumFractionDigits={3}
          maximumFractionDigits={3}
        />
        {` ${transactionCurrency}`}
      </span>
    );
  };

  handleLoadMore = () => {
    const { currentUsername } = this.props;
    const { lastActionCount } = this.state;
    const limit = lastActionCount < defaultAccountLimit ? lastActionCount : defaultAccountLimit;
    this.setState({
      loadingMoreTransactions: true,
    });
    this.props
      .getMoreUserAccountHistory(currentUsername, lastActionCount, limit)
      .then((result) => {
        const newLastActionCount = _.last(result.value.userAccountHistory).actionCount;
        if (newLastActionCount === 0) {
          this.setState({
            userHasMoreTransactions: false,
            loadingMoreTransactions: false,
          });
        } else {
          this.setState({
            lastActionCount: newLastActionCount,
            loadingMoreTransactions: false,
          });
        }
      })
      .catch(() => {
        this.setState({
          userHasMoreTransactions: false,
          loadingMoreTransactions: false,
        });
      });
  };

  render() {
    const { transactions, currentUsername, totalVestingShares, totalVestingFundSteem } = this.props;
    const { loadingMoreTransactions } = this.state;

    return (
      <div className="UserWalletTransactions">
        <ReduxInfiniteScroll
          loadMore={this.handleLoadMore}
          hasMore={this.state.userHasMoreTransactions}
          elementIsScrollable={false}
          threshold={200}
          loader={<div style={{ margin: '20px' }}><Loading /></div>}
          loadingMore={loadingMoreTransactions}
        >
          {transactions.map((transaction, index) => {
            const key = `${transaction.trx_id}${index}`;
            const transactionType = transaction.op[0];
            const transactionDetails = transaction.op[1];

            switch (transactionType) {
              case 'transfer_to_vesting':
                return (
                  <PowerUpTransaction
                    key={key}
                    amount={this.getFormattedTransactionAmount(transactionDetails.amount, 'SP')}
                    timestamp={transaction.timestamp}
                  />
                );
              case 'transfer':
                if (transactionDetails.to === currentUsername) {
                  return (
                    <ReceiveTransaction
                      key={key}
                      from={transactionDetails.from}
                      memo={transactionDetails.memo}
                      amount={this.getFormattedTransactionAmount(transactionDetails.amount)}
                      timestamp={transaction.timestamp}
                    />
                  );
                }
                return (
                  <TransferTransaction
                    key={key}
                    to={transactionDetails.to}
                    memo={transactionDetails.memo}
                    amount={this.getFormattedTransactionAmount(transactionDetails.amount)}
                    timestamp={transaction.timestamp}
                  />
                );
              case 'claim_reward_balance':
                return (
                  <ClaimReward
                    key={key}
                    timestamp={transaction.timestamp}
                    rewardSteem={transactionDetails.reward_steem}
                    rewardSbd={transactionDetails.reward_sbd}
                    rewardVests={transactionDetails.reward_vests}
                    totalVestingShares={totalVestingShares}
                    totalVestingFundSteem={totalVestingFundSteem}
                  />
                );
              case 'transfer_to_savings':
              case 'transfer_from_savings':
              case 'cancel_transfer_from_savings':
                return (
                  <SavingsTransaction
                    key={key}
                    transactionDetails={transactionDetails}
                    transactionType={transactionType}
                    amount={this.getFormattedTransactionAmount(transactionDetails.amount)}
                    timestamp={transaction.timestamp}
                  />
                );
              default:
                return null;
            }
          })}
        </ReduxInfiniteScroll>
      </div>
    );
  }
}

export default UserWalletTransactions;
