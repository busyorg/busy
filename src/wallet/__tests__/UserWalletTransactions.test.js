import React from 'react';
import { shallow } from 'enzyme';
import UserWalletTransactions from '../UserWalletTransactions';

describe('(Component) UserWalletTransactions', () => {
  describe('with default prop values', () => {
    it('renders and matches snapshot', () => {
      const props = {
        transactions: [
          {
            timestamp: '0',
            op: [
              'transfer_to_vesting',
              {
                amount: '100 STEEM',
              },
            ],
          },
          {
            timestamp: '0',
            op: [
              'transfer',
              {
                from: 'hellosteem1',
                memo: 'transfer memo',
                amount: '100 STEEM',
              },
            ],
          },
        ],
        currentUsername: 'hellosteem',
        totalVestingShares: '0',
        totalVestingFundSteem: '0',
      };
      const wrapper = shallow(<UserWalletTransactions {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
