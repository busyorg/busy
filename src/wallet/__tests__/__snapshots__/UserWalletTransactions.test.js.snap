// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) UserWalletTransactions with default prop values renders and matches snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="UserWalletTransactions"
  >
    <ReduxInfiniteScroll
      className=""
      containerHeight="100%"
      elementIsScrollable={false}
      hasMore={false}
      holderType="div"
      items={Array []}
      loadMore={[Function]}
      loader={
        <div
          style={
            Object {
              "margin": "20px",
            }
          }
        >
          <Loading
            center={true}
            style={Object {}}
          />
        </div>
      }
      loadingMore={false}
      showLoader={true}
      threshold={200}
    >
      <PowerUpTransaction
        amount={
          <span>
            <FormattedNumber
              maximumFractionDigits={3}
              minimumFractionDigits={3}
              value={100}
            />
             SP
          </span>
        }
        timestamp="0"
      />
      <TransferTransaction
        amount={
          <span>
            <FormattedNumber
              maximumFractionDigits={3}
              minimumFractionDigits={3}
              value={100}
            />
             STEEM
          </span>
        }
        memo="transfer memo"
        timestamp="0"
        to=""
      />
    </ReduxInfiniteScroll>
  </div>,
  "nodes": Array [
    <div
      className="UserWalletTransactions"
    >
      <ReduxInfiniteScroll
        className=""
        containerHeight="100%"
        elementIsScrollable={false}
        hasMore={false}
        holderType="div"
        items={Array []}
        loadMore={[Function]}
        loader={
          <div
            style={
              Object {
                "margin": "20px",
              }
            }
          >
            <Loading
              center={true}
              style={Object {}}
            />
          </div>
        }
        loadingMore={false}
        showLoader={true}
        threshold={200}
      >
        <PowerUpTransaction
          amount={
            <span>
              <FormattedNumber
                maximumFractionDigits={3}
                minimumFractionDigits={3}
                value={100}
              />
               SP
            </span>
          }
          timestamp="0"
        />
        <TransferTransaction
          amount={
            <span>
              <FormattedNumber
                maximumFractionDigits={3}
                minimumFractionDigits={3}
                value={100}
              />
               STEEM
            </span>
          }
          memo="transfer memo"
          timestamp="0"
          to=""
        />
      </ReduxInfiniteScroll>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <UserWalletTransactions
        currentUsername="hellosteem"
        totalVestingFundSteem="0"
        totalVestingShares="0"
        transactions={
          Array [
            Object {
              "op": Array [
                "transfer_to_vesting",
                Object {
                  "amount": "100 STEEM",
                },
              ],
              "timestamp": "0",
            },
            Object {
              "op": Array [
                "transfer",
                Object {
                  "amount": "100 STEEM",
                  "from": "hellosteem1",
                  "memo": "transfer memo",
                },
              ],
              "timestamp": "0",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": UserWalletTransactions {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "getFormattedTransactionAmount": [Function],
        "handleLoadMore": [Function],
        "props": Object {
          "currentUsername": "hellosteem",
          "totalVestingFundSteem": "0",
          "totalVestingShares": "0",
          "transactions": Array [
            Object {
              "op": Array [
                "transfer_to_vesting",
                Object {
                  "amount": "100 STEEM",
                },
              ],
              "timestamp": "0",
            },
            Object {
              "op": Array [
                "transfer",
                Object {
                  "amount": "100 STEEM",
                  "from": "hellosteem1",
                  "memo": "transfer memo",
                },
              ],
              "timestamp": "0",
            },
          ],
        },
        "refs": Object {},
        "state": Object {
          "lastActionCount": undefined,
          "loadingMoreTransactions": false,
          "userHasMoreTransactions": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="UserWalletTransactions"
        >
          <ReduxInfiniteScroll
            className=""
            containerHeight="100%"
            elementIsScrollable={false}
            hasMore={false}
            holderType="div"
            items={Array []}
            loadMore={[Function]}
            loader={
              <div
                style={
                  Object {
                    "margin": "20px",
                  }
                }
              >
                <Loading
                  center={true}
                  style={Object {}}
                />
              </div>
            }
            loadingMore={false}
            showLoader={true}
            threshold={200}
          >
            <PowerUpTransaction
              amount={
                <span>
                  <FormattedNumber
                    maximumFractionDigits={3}
                    minimumFractionDigits={3}
                    value={100}
                  />
                   SP
                </span>
              }
              timestamp="0"
            />
            <TransferTransaction
              amount={
                <span>
                  <FormattedNumber
                    maximumFractionDigits={3}
                    minimumFractionDigits={3}
                    value={100}
                  />
                   STEEM
                </span>
              }
              memo="transfer memo"
              timestamp="0"
              to=""
            />
          </ReduxInfiniteScroll>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="UserWalletTransactions"
        >
          <ReduxInfiniteScroll
            className=""
            containerHeight="100%"
            elementIsScrollable={false}
            hasMore={false}
            holderType="div"
            items={Array []}
            loadMore={[Function]}
            loader={
              <div
                style={
                  Object {
                    "margin": "20px",
                  }
                }
              >
                <Loading
                  center={true}
                  style={Object {}}
                />
              </div>
            }
            loadingMore={false}
            showLoader={true}
            threshold={200}
          >
            <PowerUpTransaction
              amount={
                <span>
                  <FormattedNumber
                    maximumFractionDigits={3}
                    minimumFractionDigits={3}
                    value={100}
                  />
                   SP
                </span>
              }
              timestamp="0"
            />
            <TransferTransaction
              amount={
                <span>
                  <FormattedNumber
                    maximumFractionDigits={3}
                    minimumFractionDigits={3}
                    value={100}
                  />
                   STEEM
                </span>
              }
              memo="transfer memo"
              timestamp="0"
              to=""
            />
          </ReduxInfiniteScroll>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <UserWalletTransactions
    currentUsername="hellosteem"
    totalVestingFundSteem="0"
    totalVestingShares="0"
    transactions={
      Array [
        Object {
          "op": Array [
            "transfer_to_vesting",
            Object {
              "amount": "100 STEEM",
            },
          ],
          "timestamp": "0",
        },
        Object {
          "op": Array [
            "transfer",
            Object {
              "amount": "100 STEEM",
              "from": "hellosteem1",
              "memo": "transfer memo",
            },
          ],
          "timestamp": "0",
        },
      ]
    }
  />,
}
`;
